arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArLocationDataSource.kt

```diff
@@ -0,0 +1,584 @@
+        if (lastKnownLocation != null) {
+            // reset the last known location'speed and bearing, original data no meaning anymore.
+            lastKnownLocation.speed = 0f
+            lastKnownLocation.bearing = 0f
+            setLastKnownLocation(lastKnownLocation.toEsriLocation(true))
+        }
+        startLocationProviders()
+
+        startUpdateHeading()
+    }
+
+    /**
+     * Called when the location data source is stopped, by calling [LocationDataSource.stop], or indirectly by
+     * calling [LocationDisplay.stop] on the LocationDisplay associated with this data source.
+     *
+     * This method requests that location updates stop being received from the underlying Android platform location
+     * providers.
+     *
+     * @since 100.6.0
+     */
+    override fun onStop() {
+        locationManager?.removeUpdates(internalLocationListener)
+        internalLocationListener = null
+
+        // stop update heading if it is started.
+        stopUpdateHeading()
+    }
+
+    /**
+     * Validates [minTime] and [minDistance].
+     *
+     * @since 100.6.0
+     */
+    private fun checkTimeDistanceParameters(minTime: Long, minDistance: Float) {
+        if (minTime < 0) {
+            throw IllegalArgumentException(String.format(PARAMETER_OUT_OF_BOUNDS_MSG, "minTime"))
+        }
+        minimumUpdateTime = minTime
+
+        if (minDistance < 0) {
+            throw IllegalArgumentException(
+                String.format(
+                    PARAMETER_OUT_OF_BOUNDS_MSG,
+                    "minDistance"
+                )
+            )
+        }
+        minimumUpdateDistance = minDistance
+    }
+
+    /**
+     * Select the default location providers, network will be used first if it is available.
+     *
+     * @since 100.6.0
+     */
+    private fun selectProvidersByDefault() {
+        locationManager?.let {
+            // Check if the network location service enabled
+            if (it.isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
+                selectedLocationProviders.add(LocationManager.NETWORK_PROVIDER)
+            }
+
+            // check if the GPS location service enabled,if true, use it too.
+            if (it.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
+                selectedLocationProviders.add(LocationManager.GPS_PROVIDER)
+            }
+        }
+    }
+
+    /**
+     * Called when the AutoPanMode is changed on the associated LocationDisplay, this method populates the heading
+     * [LocationDisplay.getHeading] on the associated LocationDisplay if the AutoPanMode is changed to
+     * [AutoPanMode.COMPASS_NAVIGATION].
+     *
+     * @since 100.6.0
+     */
+    override fun onAutoPanModeChanged(autoPanModeEvent: LocationDisplay.AutoPanModeChangedEvent) {
+        autoPanMode = autoPanModeEvent.autoPanMode
+        if (!isStarted) {
+            return
+        }
+    }
```

Comment:
Looks like we don't need the `autoPanMode` anywhere, so I think you can completely remove the `AutoPanModeChangedListener` implementation. Also, could you create a follow up issue in the java repo so we change the `AndroidLocationDataSource` to request heading updates for any mode, not just AutoPan? As I understood it, the other platforms are doing that in their default location data source implementation?

Reply:
I'm not sure if that's the case, we'd need to check with an iOS dev as I seem to remember @mhdostal having to implement something very similar to what I have, though I could be wrong.

Reply:
To add: I've removed the references to auto pan mode, but just worth checking the heading updates are as described on the other platforms.